src/common/post.ts

```diff
@@ -373,13 +378,88 @@ export interface CreateSourcePostModeration
 interface CreateSourcePostModerationProps {
   ctx: AuthContext;
   args: CreateSourcePostModeration;
+  options?: Partial<{
+    isMultiPost: boolean;
+    entityManager: EntityManager;
+  }>;
 }
 
+const hasDuplicatedPostBy = async (
+  con: DataSource,
+  { dedupKey, sourceId }: Partial<Record<'dedupKey' | 'sourceId', string>>,
+): Promise<boolean> => {
+  if (!dedupKey) return false;
+
+  return await queryReadReplica(con, async ({ queryRunner }) => {
+    const pendingQb = queryRunner.manager
+      .getRepository(SourcePostModeration)
+      .createQueryBuilder('p')
+      .where({
+        status: SourcePostModerationStatus.Pending,
+      })
+      .andWhere(`p.flags->> 'dedupKey' = :dedupKey`, { dedupKey });
+    const postQb = queryRunner.manager
+      .getRepository(Post)
+      .createQueryBuilder('p')
+      .where(`p.flags->> 'dedupKey' = :dedupKey`, { dedupKey });
+
+    if (sourceId) {
+      pendingQb.andWhere(`p.sourceId = :sourceId`, { sourceId });
+      postQb.andWhere(`p.sourceId = :sourceId`, { sourceId });
+    }
+
+    const [pendingExists, postExists] = await Promise.all([
+      pendingQb.getExists(),
+      postQb.getExists(),
+    ]);
+
+    return pendingExists || postExists;
+  });
+};
+
+const getModerationWarningFlag = async ({
+  con,
+  isMultiPost = false,
+  dedupKey,
+  sourceId,
+}: {
+  con: DataSource;
+  isMultiPost?: boolean;
+  dedupKey?: string;
+  sourceId?: string;
+}): Promise<WarningReason | undefined> => {
+  if (isMultiPost) {
+    return WarningReason.MultipleSquadPost;
+  }
+
+  if (!dedupKey) {
+    return;
+  }
+
+  const isDuplicatedInSameSquad = await hasDuplicatedPostBy(con, {
+    dedupKey,
+    sourceId,
+  });
+  if (isDuplicatedInSameSquad) {
+    return WarningReason.DuplicatedInSameSquad;
+  }
+
+  const isDuplicatedAcrossSquads = await hasDuplicatedPostBy(con, {
```

Comment:
why do we call `hasDuplicatedPostBy` twice, once with sourceId once without? Would it not do the same thing if sourceId is not passed thus doing the query twice?